Add fallback value support in PowerDistributingActor for PV inverters

shsms opened the pull request:
When an PV inverter is not available, its power is assumed to be the given fallback value, and this fallback power is excluded from the requested power that's set to the rest of the PV inverters.

shsms committed:
Add support for `fallback_power` in `PowerDistributingActor`

When a component is not available and its power can't be determined
from a connected meter, its power is assumed to be `fallback_power`.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Disallow fallback power for batteries and EV chargers

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Support customizable fallback power for PV inverters

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms commented:
This adds a new argument to the `microgrid.initialize` method called `pv_fallback_power`.  This will be used in the PV power formula calculation as well.

The problem is that `Power.zero()` is probably not the best default value for the field.  And until we have the new API, we don't have access to the rated bounds from the component info, so unless the component streamed some amount of data, we don't know what a max value could be for this field.

And then there is the issue of inverters with varying power ratings at the same location.

Do we keep it optional or make it mandatory?  Should we allow setting fallback power for individual PV inverters or just one average value?

Any ideas?  @cwasicki and @frequenz-floss/python-sdk-team ?

llucax reviewed (commented):
In general LGTM, but replying to your message:

> This adds a new argument to the microgrid.initialize method called pv_fallback_power. This will be used in the PV power formula calculation as well.

I wonder if this can't be set on a per-pool base instead for the whole microgrid. It seems weird to have this specific config at the microgrid level.

> The problem is that Power.zero() is probably not the best default value for the field. And until we have the new API, we don't have access to the rated bounds from the component info, so unless the component streamed some amount of data, we don't know what a max value could be for this field.
> 
> And then there is the issue of inverters with varying power ratings at the same location.

Not sure about this, if in the future we can get the rated power, maybe we can introduce a sentinel like `MAX_RATED_POWER` to say we want to set the fallback to the rated power, and maybe `MAX_CURRENT_BOUND` to follow the current bounds?

> Do we keep it optional or make it mandatory? Should we allow setting fallback power for individual PV inverters or just one average value?

I think long term it would be good to be able to have per-component fallback, again for me ideally at the pool level, so we can do something like:

```py
pv_pool1 = microgrid.new_pv_pool([1, 2], fallback_power={
    1: Power.from_kilowatt(50), 2: MAX_RATED_POWER}
)
# uses the same for all components in the pool
pv_pool2 = microgrid.new_pv_pool(fallback_power=MAX_RATED_POWER)
```

shsms commented:
> I wonder if this can't be set on a per-pool base instead for the whole microgrid. It seems weird to have this specific config at the microgrid level.

I think per-pool won't work because each pool could then have different fallback values, and then some combinations would mean that lower priority actors are able to exceed what a higher priority actor would have set for a given set of bounds.  Else we'd have to use the power manager to apply restrictions on the fallback power as well, which would get very confusing.

> maybe we can introduce a sentinel like MAX_RATED_POWER to say we want to set the fallback to the rated power

yes, with the new API this would be possible, and would simplify actor code.

> I think long term it would be good to be able to have per-component fallback

supporting everything like in your example sounds like a nice idea.

cwasicki commented:
>I wonder if this can't be set on a per-pool base instead for the whole microgrid. It seems weird to have this specific config at the microgrid level.

If this is possible I would also prefer setting it on pool level.

> And until we have the new API, we don't have access to the rated bounds from the component info

I didn't know we don't have access to these bounds yet, how far are these away?

> so unless the component streamed some amount of data, 

Is this a suggestion to keep track of the maximum value observed so far? Doesn't sound like a bad idea. We would need the same logic later just using the bounds right, since for unreachable inverters we cannot get the bounds neither.

> Should we allow setting fallback power for individual PV inverters or just one average value?

For an intermediate solution I think one value is fine, in practice we would just be conservative about it.

llucax commented:
> I think per-pool won't work because each pool could then have different fallback values, and then some combinations would mean that lower priority actors are able to exceed what a higher priority actor would have set for a given set of bounds. Else we'd have to use the power manager to apply restrictions on the fallback power as well, which would get very confusing.

I see, what a shame. I think also it makes things more complicated, because different actors will probably want to assume different outcomes, so when we have one app combining multiple actors this might be problematic. For example PV optimization might want to assume the missing power is MAX to avoid grid feed-in, but peak shaving might want to assume missing power is MIN to avoid peaks. :thinking:

shsms commented:
> > And until we have the new API, we don't have access to the rated bounds from the component info
> 
> I didn't know we don't have access to these bounds yet, how far are these away?

rated bounds are only available in the data streams in the current version of the api.  In the upcoming version, we'll be able to get rated bounds even if there's no live data available, since the start of the app.

> 
> > so unless the component streamed some amount of data,
> 
> Is this a suggestion to keep track of the maximum value observed so far? Doesn't sound like a bad idea. We would need the same logic later just using the bounds right, since for unreachable inverters we cannot get the bounds neither.

no, if we have received any streaming data, we'll have the rated bounds.  so if components were up during application startup and then go down, then the app will have data.  If the app has to restart when a component is unreachable, then we'll not have the info.

ela-kotulska-frequenz commented on src/frequenz/sdk/microgrid/_power_distributing/_component_managers/_battery_manager.py:
```diff
@@ -151,6 +151,10 @@ def __init__(
         self._api_power_request_timeout = api_power_request_timeout
         self._fallback_power = fallback_power
 
+        assert (
+            self._fallback_power == Power.zero()
+        ), "Non-zero fallback power is not supported for batteries."
+
```
Maybe you could mention this in the `fallback_power` docs 4 lines above?
It only says:
```
            fallback_power: The power to assume a battery has if the battery is not
                reachable.
```

shsms replied:
Done.

ela-kotulska-frequenz commented on src/frequenz/sdk/microgrid/_power_distributing/_component_managers/_ev_charger_manager/_ev_charger_manager.py:
```diff
@@ -66,6 +66,10 @@ def __init__(
+        assert (
+            self._fallback_power == Power.zero()
+        ), "Non-zero fallback power is not supported for EV chargers."
+
```
Same, maybe you could mention this in the `fallback_power` docs 4 lines above?
It only says:
```
            fallback_power: The power to assume a battery has if the battery is not
                reachable.
```

shsms replied:
Updated.

ela-kotulska-frequenz commented on src/frequenz/sdk/microgrid/_power_distributing/_component_managers/_pv_inverter_manager/_pv_inverter_manager.py:
```diff
@@ -263,3 +284,69 @@ def _get_pv_inverter_ids(self) -> collections.abc.Set[int]:
+        Returns:
+            The power of the unreachable inverters in the request.
+        """
+        unreachable_inv_ids = request.component_ids - set(working_components)
+        if not unreachable_inv_ids:
+            return Power.zero()
+        unreachable_power = Power.zero()
+        now = datetime.now(tz=timezone.utc)
+        for inv_id in unreachable_inv_ids:
+            if inv_id not in self._connected_meters:
+                unreachable_power += self._fallback_power
+                continue
+
+            meter_id = self._connected_meters[inv_id]
+            meter_cache = self._meter_data_caches[meter_id]
+            if not meter_cache.has_value():
+                unreachable_power += self._fallback_power
+                continue
+
+            meter_data = meter_cache.get()
+            if now - meter_data.timestamp > MAX_DATA_AGE:
+                unreachable_power += self._fallback_power
+                continue
+
+            unreachable_power = Power.from_watts(
+                unreachable_power.as_watts() + meter_data.active_power
+            )
+        return unreachable_power
```
Not sure but from logs I saw components are changing state from not-working to working after  ~10sec . 
That would mean that every 10 sec we would try to distribute whole requested pv power - and (probably) feed -in.

Maybe we should check last message timestamp for all meters connected to the requested inverters - not only the unreachable one ?

shsms replied:
If components are not sending data, their state doesn't change to `WORKING` until there's data again.  Automatic change only happens from `UNCERTAIN` to `WORKING`, never from `NOT_WORKING` to `WORKING`.  (And `UNCERTAIN` happens when there's data and no error state in the data, but a request to the API has failed, and `NOT_WORKING` happens 10 seconds after there's no data or immediately after data shows component is in error state.)

I've also tested this code and that's not an issue.

ela-kotulska-frequenz replied:
Thanks

shsms committed:
Switch to a custom component config for PVPool control tests

This config allows the tests to run in scenarios where there are
multiple inverters attached to the same meter.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add tests for `pv_fallback_power` in power distribution

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms commented:
Will try to get the formula side done as well, before asking for a review again.

shsms commented:
We have decided to not go ahead with this approach and instead stick to what was originally planned in https://github.com/frequenz-floss/frequenz-sdk-python/issues/1024
